src/main/scala/viper/gobra/frontend/Config.scala

```diff
@@ -370,76 +475,88 @@ class ScallopGobraConfig(arguments: Seq[String], isInputOptional: Boolean = fals
 
   private object InputConverter {
 
-    def validate(input: List[String]): Messages = {
-      val files = input map isGoFilePath
-      files.partition(_.isLeft) match {
-        case (pkgs,  files) if pkgs.length == 1 && files.isEmpty => noMessages
-        case (pkgs, files) if pkgs.isEmpty && files.nonEmpty => noMessages
-        // error states:
-        case (pkgs,  files) if pkgs.length > 1 && files.isEmpty =>
-          message(pkgs, s"multiple package names provided: '${concatLeft(pkgs, ",")}'")
-        case (pkgs, files) if pkgs.nonEmpty && files.nonEmpty =>
-          message(pkgs, s"specific input files and one or more package names were simultaneously provided (files: '${concatRight(files, ",")}'; package names: '${concatLeft(pkgs, ",")}')")
+    def validate(inputs: List[String], recursive: Boolean): Messages = {
+      val files = inputs flatMap (file => getAllGobraFiles(file, recursive))
+      files match {
+        case files if files.nonEmpty => noMessages
         case _ if isInputOptional => noMessages
-        case (pkgs, files) if pkgs.isEmpty && files.isEmpty => message(null, s"no input specified")
+        // error states
+        case files if files.isEmpty => message(null, s"no files found in the specified input (use -r to traverse directories)")
         case c => Violation.violation(s"This case should be unreachable, but got $c")
       }
     }
 
-    def convert(input: List[String], moduleName: String, includeDirs: Vector[Path]): Vector[Source] = {
-      val res = for {
-        i <- identifyInput(input).toRight("invalid input")
-        sources <- i match {
-          case Right(filePaths) => Right(filePaths.map(f => FileSource(f.toString)))
-          case Left(_) => PackageResolver.resolveSources(RegularImport(""), moduleName, includeDirs)
-        }
-      } yield sources
-      assert(isInputOptional || res.isRight, s"validate function did not catch this problem: '${res.swap.getOrElse(None)}'")
-      res.getOrElse(Vector())
-    }
+    def convert(input: List[String], includeDirs: Vector[Path], recursive: Boolean, includePackages: List[String], excludePackages: List[String]): Map[String, Vector[Source]] = {
```

Comment:
do you retrieve all files recursively contains in the specified folders and then group them by package name?
What happens in the case where files in different directories have the same package clause? Based on my understanding of the Go spec, the package clause does not need to be unique and there can be multiple directories that have source files specifying the same package clause